### ISSUE TITLE
ERROR: 422 on query with binary operation and keep_metric_names

### ISSUE THREAD

**Issue by tekert** [VictoriaMetrics/VictoriaMetrics#6259]
### Describe the bug

```
rate({__name__=~"node_netstat_Ip(6|)_(Reasm|Frag).*", job="node_exporter"} * 30) keep_metric_names
ERROR: 422, error when executing query="rate({__name__=~\"node_netstat_Ip(6|)_(Reasm|Frag).*\", job=\"node_exporter\"} * 1) keep_metric_names" on the time range (start=1715589501789, end=1715590401789, step=30000): duplicate output timeseries: {group="routers", instance="<redacted>", job="node_exporter"}
```

### To Reproduce

30s is the scrape interval.

If i remove the * 30 then it works.

If i ask for a interval like:

`rate({__name__=~"node_netstat_Ip(6|)_(Reasm|Frag).*", job="node_exporter"}[30s] * 30) keep_metric_names`
Same error.

There are 12 metrics that match that regex, half return 0 always.

### Version

victoria-metrics-windows-amd64-v1.101.0

**Comment by zekker6** [comment]
Hello @tekert 

This is caused by [implicit query conversion](https://docs.victoriametrics.com/metricsql/#implicit-query-conversions):
> If something other than [series selector](https://docs.victoriametrics.com/keyconcepts/#filtering) is passed to [rollup function](https://docs.victoriametrics.com/metricsql/#rollup-functions), then a [subquery](https://docs.victoriametrics.com/metricsql/#subqueries) with 1i lookbehind window and 1i step is automatically formed. For example, rate(sum(up)) is automatically converted to rate((sum(default_rollup(up)))[1i:1i]). This behavior can be disabled or logged via cmd-line flags -search.disableImplicitConversion and -search.logImplicitConversion since v1.101.0.

So `{__name__=~"node_netstat_Ip(6|)_(Reasm|Frag).*", job="node_exporter"} * 30` is evaluated as a subquery. Also, this will be further transformed into `default_rollup({__name__=~"node_netstat_Ip(6|)_(Reasm|Frag).*", job="node_exporter"}) + 30`.
In order to make that work as you expect you need to use the following query: `rate({__name__=~"node_netstat_Ip(6|)_(Reasm|Frag).*", job="node_exporter"} * 30 keep_metric_names) keep_metric_names`.
See [this ](https://play.victoriametrics.com/select/0/vmui/?#/?g0.expr=rate%28%7B__name__%3D%7E%22node_netstat_Ip%286%7C%29_%28Out%7CIn%29.*%22%7D+*+30+keep_metric_names%29+keep_metric_names&g0.range_input=30m&g0.end_input=2024-05-13T10%3A07%3A24&g0.tab=0&g0.relative_time=last_30_minutes&g0.tenantID=0) for example.


Also, starting from [v1.101.0](https://github.com/VictoriaMetrics/VictoriaMetrics/releases/tag/v1.101.0) it is possible to either enable logging for implicit query conversions or disable this completely by using `-search.logImplicitConversion` or  `-search.disableImplicitConversion` command-line flags respectively.

**Comment by tekert** [comment]
Thanks @zekker6 , "other than series selector" mean a subquery that returns multiple metrics in this case?
 
So to understand correctly if i disable implicit conversion, what would i have to write? if I write the query without * 30 :
`rate(default_rollup({__name__=~"node_netstat_Ip(6|)_(Out|In).*"})[30s]) keep_metric_names`
Works, and to add a binary operation, seems i have to do it inside the selection like this
`rate(default_rollup({__name__=~"node_netstat_Ip(6|)_(Out|In).*"} * 30 keep_metric_names)) keep_metric_names`
And not like this:
`rate(default_rollup({__name__=~"node_netstat_Ip(6|)_(Out|In).*"}) * 30 keep_metric_names) keep_metric_names`

**Comment by zekker6** [comment]
@tekert 
> "other than series selector" mean a subquery that returns multiple metrics in this case?

No, in this case `rate()` takes a binary operation expression as an argument(`a * b`) which leads to using implicit convertion. 

> So to understand correctly if i disable implicit conversion, what would i have to write? if I write the query without * 30 :
> edit: funny if i multiply doing * 30s (s at the end) it works too, wonder what is doing.

The idea is to pass a metric selector to the rollup function, so when you're using `rate(a) * b`  there will be no subquery since rate argument is a metric selector.

**Comment by tekert** [comment]
I see, thanks @zekker6  closing this.
